algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs

```diff
@@ -25,34 +26,32 @@ use crate::{
     },
 };
 
+use itertools::Itertools;
 use rand_core::RngCore;
 use snarkvm_fields::PrimeField;
 
 impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+    /// Output the number of oracles sent by the prover in the third round.
+    pub fn num_fourth_round_oracles() -> usize {
+        1
+    }
+
     /// Output the fourth round message and the next state.
     pub fn prover_fourth_round<R: RngCore>(
-        verifier_message: &verifier::ThirdMessage<F>,
-        state: prover::State<F, MM>,
+        verifier_message: &mut verifier::ThirdMessage<F>,
+        mut state: prover::State<'_, F, MM>,
         _r: &mut R,
     ) -> Result<prover::FourthOracles<F>, AHPError> {
-        let verifier::ThirdMessage { r_b, r_c, .. } = verifier_message;
-        let [mut lhs_a, mut lhs_b, mut lhs_c] = state.lhs_polynomials.unwrap();
-        lhs_b *= *r_b;
-        lhs_c *= *r_c;
-
-        lhs_a += &lhs_b;
-        lhs_a += &lhs_c;
-        let h_2 = LabeledPolynomial::new("h_2".into(), lhs_a, None, None);
+        let mut lhs_sum = DensePolynomial::zero();
+        for (&r, lhs) in verifier_message.iter().zip_eq(state.lhs_polys_iter_mut()) {
+            lhs_sum += &(std::mem::take(lhs) * r);
```

Comment:
I think for this we can break it up into two steps: multiply lhs in place by r (`lhs *= r`), and then `lhs_sum += std::mem::take(lhs);`

Also, it might be a good idea to parallelize this with a `fold`